File: src/main/java/com/pedigree/model/Family.java
package com.pedigree.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class Family {
    private final String id;
    private String husbandId;
    private String wifeId;
    private final List<String> childrenIds;
    private String marriageDate;
    private String marriagePlace;
    private final List<GedcomEvent> events; // All family events (MARR, DIV, ANUL, etc.)
    private final List<SourceCitation> sources; // Source citations
    private final List<Note> notes;
    private final List<MediaAttachment> media;
    private final List<Tag> tags;

    public Family() {
        this(UUID.randomUUID().toString());
    }

    public Family(String id) {
        this.id = Objects.requireNonNull(id, "id");
        this.childrenIds = new com.pedigree.util.DirtyObservableList<>();
        this.events = new com.pedigree.util.DirtyObservableList<>();
        this.sources = new com.pedigree.util.DirtyObservableList<>();
        this.notes = new com.pedigree.util.DirtyObservableList<>();
        this.media = new com.pedigree.util.DirtyObservableList<>();
        this.tags = new com.pedigree.util.DirtyObservableList<>();
    }

    @JsonCreator
    public Family(
            @JsonProperty("id") String id,
            @JsonProperty("husbandId") String husbandId,
            @JsonProperty("wifeId") String wifeId,
            @JsonProperty("childrenIds") List<String> childrenIds,
            @JsonProperty("marriageDate") String marriageDate,
            @JsonProperty("marriagePlace") String marriagePlace,
            @JsonProperty("marriage") Event marriage,
            @JsonProperty("events") List<GedcomEvent> events,
            @JsonProperty("sources") List<SourceCitation> sources,
            @JsonProperty("notes") List<Note> notes,
            @JsonProperty("media") List<MediaAttachment> media,
            @JsonProperty("tags") List<Tag> tags
    ) {
        this.id = Objects.requireNonNull(id, "id");
        this.husbandId = husbandId;
        this.wifeId = wifeId;
        this.childrenIds = (childrenIds != null) ? new com.pedigree.util.DirtyObservableList<>(childrenIds) : new com.pedigree.util.DirtyObservableList<>();
        // Support backward compatibility: if old Event marriage exists, convert it
        if (marriage != null) {
            this.marriageDate = marriage.getDate() != null ? marriage.getDate().toString() : null;
            this.marriagePlace = marriage.getPlace();
        } else {
            this.marriageDate = marriageDate;
            this.marriagePlace = marriagePlace;
        }
        this.events = (events != null) ? new com.pedigree.util.DirtyObservableList<>(events) : new com.pedigree.util.DirtyObservableList<>();
        this.sources = (sources != null) ? new com.pedigree.util.DirtyObservableList<>(sources) : new com.pedigree.util.DirtyObservableList<>();
        this.notes = (notes != null) ? new com.pedigree.util.DirtyObservableList<>(notes) : new com.pedigree.util.DirtyObservableList<>();
        this.media = (media != null) ? new com.pedigree.util.DirtyObservableList<>(media) : new com.pedigree.util.DirtyObservableList<>();
        this.tags = (tags != null) ? new com.pedigree.util.DirtyObservableList<>(tags) : new com.pedigree.util.DirtyObservableList<>();
    }

    public String getId() { return id; }

    @JsonProperty("husbandId")
    public String getHusbandId() { return husbandId; }

    public void setHusbandId(String husbandId) {
        this.husbandId = husbandId;
        com.pedigree.util.DirtyFlag.setModified();
    }

    @JsonProperty("wifeId")
    public String getWifeId() { return wifeId; }

    public void setWifeId(String wifeId) {
        this.wifeId = wifeId;
        com.pedigree.util.DirtyFlag.setModified();
    }

    public List<String> getChildrenIds() { return childrenIds; }

    public String getMarriageDate() { return marriageDate; }

    public void setMarriageDate(String marriageDate) {
        this.marriageDate = marriageDate;
        com.pedigree.util.DirtyFlag.setModified();
    }

    public String getMarriagePlace() { return marriagePlace; }

    public void setMarriagePlace(String marriagePlace) {
        this.marriagePlace = marriagePlace;
        com.pedigree.util.DirtyFlag.setModified();
    }

    public List<GedcomEvent> getEvents() { return events; }

    public List<SourceCitation> getSources() { return sources; }

    public List<Note> getNotes() { return notes; }

    public List<MediaAttachment> getMedia() { return media; }

    public List<Tag> getTags() { return tags; }
}

File: src/main/kotlin/com/family/tree/model/Family.kt
package com.family.tree.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.family.tree.util.DirtyFlag
import com.family.tree.util.DirtyObservableList
import java.util.Objects
import java.util.UUID

@JsonIgnoreProperties(ignoreUnknown = true)
class Family {
    val id: String
    var husbandId: String? = null
        @JsonProperty("husbandId")
        set(value) { field = value; DirtyFlag.setModified() }
    var wifeId: String? = null
        @JsonProperty("wifeId")
        set(value) { field = value; DirtyFlag.setModified() }
    val childrenIds: MutableList<String>
    val events: MutableList<GedcomEvent>
    val sources: MutableList<SourceCitation>
    val notes: MutableList<Note>
    val media: MutableList<MediaAttachment>
    val tags: MutableList<Tag>

    constructor() : this(UUID.randomUUID().toString())

    constructor(id: String) {
        this.id = Objects.requireNonNull(id, "id")
        this.childrenIds = DirtyObservableList()
        this.events = DirtyObservableList()
        this.sources = DirtyObservableList()
        this.notes = DirtyObservableList()
        this.media = DirtyObservableList()
        this.tags = DirtyObservableList()
    }

    @JsonCreator
    constructor(
        @JsonProperty("id") id: String?,
        @JsonProperty("husbandId") husbandId: String?,
        @JsonProperty("wifeId") wifeId: String?,
        @JsonProperty("childrenIds") childrenIds: List<String>?,
        @JsonProperty("events") events: List<GedcomEvent>?,
        @JsonProperty("sources") sources: List<SourceCitation>?,
        @JsonProperty("notes") notes: List<Note>?,
        @JsonProperty("media") media: List<MediaAttachment>?,
        @JsonProperty("tags") tags: List<Tag>?
    ) {
        this.id = id ?: throw NullPointerException("id")
        this.husbandId = husbandId
        this.wifeId = wifeId
        this.childrenIds = if (childrenIds != null) DirtyObservableList(childrenIds) else DirtyObservableList()
        this.events = if (events != null) DirtyObservableList(events) else DirtyObservableList()
        this.sources = if (sources != null) DirtyObservableList(sources) else DirtyObservableList()
        this.notes = if (notes != null) DirtyObservableList(notes) else DirtyObservableList()
        this.media = if (media != null) DirtyObservableList(media) else DirtyObservableList()
        this.tags = if (tags != null) DirtyObservableList(tags) else DirtyObservableList()
    }
}
